### PR TITLE
add nameconstraints classes

### PR DESCRIPTION
**Pull request by reaperhulk**
refs #1947

### CODE REVIEW

**Comment by reaperhulk** [comment]
...and I forgot docs. Expect those later (__eq__ will be in a future PR though to keep this one smaller)

**Comment by reaperhulk** [comment]
Okay, docs are pushed now. I stripped down the `GeneralSubtree` class because I discovered these lines in RFC 5280:

```
Within this profile, the minimum and maximum fields are not used with
   any name forms, thus, the minimum MUST be zero, and maximum MUST be
   absent
```

Since that seems pretty foolish I decided to just drop them. Since we still have the object if it turns out there's a separate profile that uses them we can always add them back later.

**Comment by alex** [comment]
needs a rebase

**Comment by reaperhulk** [comment]
The `GeneralSubtree` object is a bit annoying. In the RFC it is defined as having 3 properties: `base`, `minimum`, and `maximum`. The spec has this to say about them:

```
Within this profile, the minimum and maximum fields are not used with
   any name forms, thus, the minimum MUST be zero, and maximum MUST be
   absent.  However, if an application encounters a critical name
   constraints extension that specifies other values for minimum or
   maximum for a name form that appears in a subsequent certificate, the
   application MUST either process these fields or reject the
   certificate
```

So the presence of this object is arguably future-proofing in case we need to add minimum/maximum support in the future.

Alternately, we could simplify this PR by removing the `GeneralSubtree` entirely and just have the permitted_subtrees and excluded_subtrees be a list of GeneralName objects.

File: docs/x509.rst [hunk 1]
```diff
@@ -735,6 +735,50 @@ X.509 Extensions
...
+    the way this extension should be processed see :rfc:`5280`.
+
+    .. attribute:: permitted_subtrees
+
+        :type: list of :class:`GeneralSubtree` objects or None
+
+        The set of permitted name patterns. If a name matches this and an
+        element in ``excluded_subtrees`` it is invalid. ``permitted_subtrees``
+        and ``excluded_subtrees``. At least one of ``permitted_subtrees`` and
+        ``excluded_subtrees`` will be non-None.
```
Comment (alex) [inline docs/x509.rst]: These seems to be an extra fragment in here?

File: docs/x509.rst [hunk 2]
```diff
@@ -735,6 +735,50 @@ X.509 Extensions
...
+    .. versionadded:: 1.0
+
+    A helper object used in :class:`NameConstraints`.
+
+    .. attribute:: base
+
+        :type: :class:`GeneralName`
+
+        The name used as the base name pattern for permitted or excluded
+        matching.
```
Comment (alex) [inline docs/x509.rst]: What value does this object add above just putting a `GeneralName` in the lists on `NameConstraints`?

File: docs/x509.rst [hunk 3]
```diff
@@ -735,6 +735,50 @@ X.509 Extensions
...
+    .. versionadded:: 1.0
+
+    The name constraints extension, which must be used only in a CA
+    certificate, indicates a name space within which all subject names in
+    subsequent certificates in a certification path must be located.
+    Restrictions apply to the subject distinguished name and apply to
+    subject alternative names.  Restrictions apply only when the
+    specified name form is present.  If no name of the type is in the
+    certificate, the certificate is acceptable. For additional details on
+    the way this extension should be processed see :rfc:`5280`.
```
Comment (alex) [inline docs/x509.rst]: This whole paragraph should be rewritten for humans please :-)

File: docs/x509.rst (hunk 2 above)
Comment (reaperhulk) [inline docs/x509.rst]: This is addressed in a comment on the main PR: https://github.com/pyca/cryptography/pull/1974#issuecomment-109814935

I'm leaning towards saying we should drop the GeneralSubtree class, but interested in your take.

Comment (alex) [inline docs/x509.rst]: How long has this RFC been a thing?

On Sun, Jun 21, 2015 at 4:56 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1974#discussion_r32897518>:
>
> > +        invalid regardless of information appearing in the
> > +        ``permitted_subtrees``. At least one of ``permitted_subtrees`` and
> > +        ``excluded_subtrees`` will be non-None.
> > +
> > +.. class:: GeneralSubtree
> > +
> > +    .. versionadded:: 1.0
> > +
> > +    A helper object used in :class:`NameConstraints`.
> > +
> > +    .. attribute:: base
> > +
> > +        :type: :class:`GeneralName`
> > +
> > +        The name used as the base name pattern for permitted or excluded
> > +        matching.
>
> This is addressed in a comment on the main PR: #1974 (comment)
> <https://github.com/pyca/cryptography/pull/1974#issuecomment-109814935>
>
> I'm leaning towards saying we should drop the GeneralSubtree class, but
> interested in your take.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1974/files#r32897518>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (reaperhulk) [inline docs/x509.rst]: RFC 5280 was finalized May 2008, so 7 years. It is really just an update to RFC 3280 (April 2002) though (and 3280 includes the same language regarding the minimum/maximum fields)

Comment (alex) [inline docs/x509.rst]: So, 13 years. Have you ever seen a cert anywhere that used these, or a
conversation anywhere about giving them meaning?

On Sun, Jun 21, 2015 at 5:00 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1974#discussion_r32897566>:
>
> > +        invalid regardless of information appearing in the
> > +        ``permitted_subtrees``. At least one of ``permitted_subtrees`` and
> > +        ``excluded_subtrees`` will be non-None.
> > +
> > +.. class:: GeneralSubtree
> > +
> > +    .. versionadded:: 1.0
> > +
> > +    A helper object used in :class:`NameConstraints`.
> > +
> > +    .. attribute:: base
> > +
> > +        :type: :class:`GeneralName`
> > +
> > +        The name used as the base name pattern for permitted or excluded
> > +        matching.
>
> RFC 5280 was finalized May 2008, so 7 years. It is really just an update
> to RFC 3280 (April 2002) though (and 3280 includes the same language
> regarding the minimum/maximum fields)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1974/files#r32897566>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

Comment (reaperhulk) [inline docs/x509.rst]: Nope.

Comment (alex) [inline docs/x509.rst]: Yeah. Kill it. We can do a deprecation dance if anything changes here.

On Sun, Jun 21, 2015 at 5:11 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1974#discussion_r32897668>:
>
> > +        invalid regardless of information appearing in the
> > +        ``permitted_subtrees``. At least one of ``permitted_subtrees`` and
> > +        ``excluded_subtrees`` will be non-None.
> > +
> > +.. class:: GeneralSubtree
> > +
> > +    .. versionadded:: 1.0
> > +
> > +    A helper object used in :class:`NameConstraints`.
> > +
> > +    .. attribute:: base
> > +
> > +        :type: :class:`GeneralName`
> > +
> > +        The name used as the base name pattern for permitted or excluded
> > +        matching.
>
> Nope.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1974/files#r32897668>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: tests/test_x509_ext.py [hunk 4]
```diff
@@ -1905,6 +1905,74 @@ def test_aki_no_keyid(self, backend):
...
+        assert nc.excluded_subtrees is None
+
+    def test_repr(self):
+        permitted = [x509.DNSName(u"name.local"), x509.DNSName(u"name2.local")]
+        nc = x509.NameConstraints(
+            permitted_subtrees=permitted,
+            excluded_subtrees=None
+        )
+        assert repr(nc) == (
+            "<NameConstraints(permitted_subtrees=[<DNSName(value=name.local)>,"
```
Comment (alex) [inline tests/test_x509_ext.py]: Missing a space after the comma at the end here

Comment (reaperhulk) [inline tests/test_x509_ext.py]: Oops, fixed.

File: src/cryptography/x509.py [hunk 5]
```diff
@@ -676,6 +676,58 @@ def __ne__(self, other):
...
+                isinstance(x, GeneralName) for x in excluded_subtrees
+            ):
+                raise TypeError(
+                    "excluded_subtrees must be a list of GeneralName objects "
+                    "or None"
+                )
+
+            self._validate_ip_name(excluded_subtrees)
+
+        if not permitted_subtrees and not excluded_subtrees:
```
Comment (alex) [inline src/cryptography/x509.py]: all the bool checks on these should use `is None`, and not the implicit conversion

**Pull request merged**